Did you try to run unit tests? They seems broken


 

<details>

<summary>log</summary>

```

======================================================================
ERROR: test_startSegments (test.unit.test_unit_gpsegstart.GpSegStart)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/test/unit/test_unit_gpsegstart.py", line 51, in test_startSegments
    exitCode = gpsegstart.run()
  File "/home/gpadmin/gpdb_src/gpMgmt/sbin/gpsegstart.py", line 301, in run
    self.__startSegments()
  File "/home/gpadmin/gpdb_src/gpMgmt/sbin/gpsegstart.py", line 243, in __startSegments
    if res.rc != 0:
AttributeError: 'NoneType' object has no attribute 'rc'

======================================================================
ERROR: test_case_5 (test.unit.test_unit_gpversion.GpVersionTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/gpversion.py", line 119, in __init__
    raise Exception("too many tokens in version")
Exception: too many tokens in version

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/test/unit/test_unit_gpversion.py", line 57, in test_case_5
    vShort = GpVersion("postgres (Greenplum Database) 3.3.5.0 build 3")
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/gpversion.py", line 183, in __init__
    raise Exception("Unrecognised Greengage Version '%s' due to %s" %
Exception: Unrecognised Greengage Version 'postgres (Greenplum Database) 3.3.5.0 build 3' due to too many tokens in version

======================================================================
ERROR: test_case_6 (test.unit.test_unit_gpversion.GpVersionTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/gpversion.py", line 119, in __init__
    raise Exception("too many tokens in version")
Exception: too many tokens in version

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/test/unit/test_unit_gpversion.py", line 69, in test_case_6
    vLong = GpVersion("PostgreSQL 8.2.14 (Greenplum Database 3.4.filerep build 0) on i386-apple-darwin9.8.0, compiled by GCC i686-apple-darwin9-gcc-4.0.1 (GCC) 4.0.1 (Apple Inc. build 5465) compiled on Feb 16 2010 11:25:31 (with assert checking)")
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/gpversion.py", line 183, in __init__
    raise Exception("Unrecognised Greengage Version '%s' due to %s" %
Exception: Unrecognised Greengage Version 'PostgreSQL 8.2.14 (Greenplum Database 3.4.filerep build 0) on i386-apple-darwin9.8.0, compiled by GCC i686-apple-darwin9-gcc-4.0.1 (GCC) 4.0.1 (Apple Inc. build 5465) compiled on Feb 16 2010 11:25:31 (with assert checking)' due to too many tokens in version

======================================================================
FAIL: test_initFromCatalog_mismatched_versions (test.unit.test_unit_gparray.GpArrayTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/gpversion.py", line 119, in __init__
    raise Exception("too many tokens in version")
Exception: too many tokens in version

During handling of the above exception, another exception occurred:

Exception: Unrecognised Greengage Version 'PostgreSQL 8.3.23 (Greenplum Database 5.0.0 build dev) on x86_64-pc-linux-gnu, compiled by GCC gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-17) compiled on Feb  9 2017 23:06:31' due to too many tokens in version

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/mock/mock.py", line 1400, in patched
    return func(*newargs, **newkeywargs)
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/test/unit/test_unit_gparray.py", line 235, in test_initFromCatalog_mismatched_versions
    with self.assertRaisesRegex(Exception, 'Cannot connect to GPDB version 5 from installed version 7'):
AssertionError: "Cannot connect to GPDB version 5 from installed version 7" does not match "Unrecognised Greengage Version 'PostgreSQL 8.3.23 (Greenplum Database 5.0.0 build dev) on x86_64-pc-linux-gnu, compiled by GCC gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-17) compiled on Feb  9 2017 23:06:31' due to too many tokens in version"

======================================================================
FAIL: test_truncate_batch_size (test.unit.test_unit_gpcheckcat.GpCheckCatTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/mock/mock.py", line 1400, in patched
    return func(*newargs, **newkeywargs)
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckcat.py", line 164, in test_truncate_batch_size
    self.assertEqual(last_call, "Truncated batch size to number of primaries: 50")
AssertionError: 'Error: only Greengage database version >= 6.0 are supported\n' != 'Truncated batch size to number of primaries: 50'
- Error: only Greengage database version >= 6.0 are supported
+ Truncated batch size to number of primaries: 50

======================================================================
FAIL: test_get_current_logfile_path (test.unit.test_unit_recoveryinfo.RecoveryResultTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gpadmin/gpdb_src/gpMgmt/bin/gppylib/test/unit/test_unit_recoveryinfo.py", line 545, in test_get_current_logfile_path
    self.assertIsNotNone(host)
AssertionError: unexpectedly None

----------------------------------------------------------------------
Ran 647 tests in 18.323s

FAILED (failures=3, errors=3)
make: *** [Makefile:111: unitdevel] Error 1
make: Leaving directory '/home/gpadmin/gpdb_src/gpMgmt/bin'

```

</details>

--- a/gpMgmt/bin/ggrebalance
+++ b/gpMgmt/bin/ggrebalance
@@ -98,6 +98,7 @@ def parseargs():
                       help='Allows to place primary and mirror at the same host during rebalance.')
     parser.add_option('-q', '--quiet', action='store_true', default=False,
                       help='Run in quiet mode.')
+    parser.add_option('--seed', dest="solver_seed", type="int", help='seed value for stable plan generation')
 
     # Parse the command line arguments
     options, args = parser.parse_args()
@@ -156,7 +157,7 @@ def validate_options(options, args, parser):
             'target_segment_count', 'mirror_mode', 'skip_rebalance', 'show_plan',
             'skip_resource_estimation', 'analyze', 'replay_lag', 'hba_hostnames', 'inplace_swap_roles',
             'duration', 'end',
-            'detailed_progress', 'simple_progress', 'no_progress'
+            'detailed_progress', 'simple_progress', 'no_progress', 'seed'
         ]),
 
         ('rollback_required', [
@@ -164,7 +165,7 @@ def validate_options(options, args, parser):
             'remove_hosts', 'remove_hosts_file', 'target_datadirs', 'target_datadirs_file',
             'target_segment_count', 'mirror_mode', 'skip_rebalance', 'show_plan',
             'skip_resource_estimation', 'inplace_swap_roles',
-            'detailed_progress', 'simple_progress', 'no_progress'
+            'detailed_progress', 'simple_progress', 'no_progress', 'seed'
         ]),
 
         ('detailed_progress', ['simple_progress', 'no_progress']),

--- a/gpMgmt/bin/ggrebalance
+++ b/gpMgmt/bin/ggrebalance
@@ -138,6 +138,10 @@ def validate_options(options, args, parser):
         logger.error(
             'Invalid value of --target-segment-count. It must be >= 1 and <= %d' % MAX_SEGMENTS)
         parser.exit(1)
+    
+    if options.solver_seed and options.solver_seed < 0:
+        logger.error('Invalid value of --seed. Must be >= 0')
+        parser.exit(1)
 
     incompatibility_rules = [
         # (condition_option, incompatible_options)

--- a/gpMgmt/bin/ggrebalance_modules/planner.py
+++ b/gpMgmt/bin/ggrebalance_modules/planner.py
@@ -69,8 +69,8 @@ class LogicalMove:
         return (
             f"Move Segment(content={self.seg.getSegmentContentId()}, dbid={self.seg.getSegmentDbId()}, "
             f"role={self.seg.role}){size_str}\n"
-            f"      From: {src_host}:{src_port} → {src_datadir}\n"
-            f"      To:   {dst_host}:{dst_port} → {dst_datadir}"
+            f"      From: {src_host}:{src_port}:{src_datadir}\n"
+            f"      To:   {dst_host}:{dst_port}:{dst_datadir}"
         )
     
 
@@ -772,7 +772,10 @@ class Planner:
                                                                          self.resolver)
         id_to_host = {v: k for k, v in host_mapping.items()}
         self.logger.info("Planning rebalance moves. Can take up to 60s.")
-        planning_seed_value = int.from_bytes(os.urandom(16) , 'big')
+        if self.options.solver_seed:
+            planning_seed_value = self.options.solver_seed
+        else:
+            planning_seed_value = int.from_bytes(os.urandom(16) , 'big')
         self.logger.info(f"Running randomized plan improvement with seed:{planning_seed_value}")
         lns_config = LNSConfig.from_parent(config, MAX_SOLVER_ITERS, SOLVER_TIMEOUT_SEC)
         solution, _ = LNS(lns_config, seed=planning_seed_value).solve()

--- a/gpMgmt/bin/ggrebalance_modules/solver.py
+++ b/gpMgmt/bin/ggrebalance_modules/solver.py
@@ -1183,7 +1183,6 @@ class LNS:
         self.solver = GreedySolver(config)
         self.rng = random.Random(seed)
         self.n_segments = config.n_segments
-        self.strategy = config.strategy
         self.printing = printing
 
         self._destroy = LNSDestroyOperators(

--- a/gpMgmt/bin/ggrebalance_modules/solver.py
+++ b/gpMgmt/bin/ggrebalance_modules/solver.py
@@ -19,10 +19,6 @@ Solution = Dict[ContentId, Tuple[HostId, HostId]]
 def _pack_solution(p: List[int], m: List[int], n: int) -> Solution:
     return {ContentId(i): (HostId(p[i]), HostId(m[i])) for i in range(n)}
 
-def _extract_mapping(sol: Solution, n: int) -> Tuple[List[int], List[int]]:
-    return [sol[i][0] for i in range(n)], [sol[i][1] for i in range(n)]
-
-
 @dataclass
 class SolverConfig:
     """
@@ -111,10 +107,7 @@ class GreedySolver:
         # Phase 2: Assign mirrors
         mirror_mapping = self._assign_mirror_hosts(primary_mapping)
                    
-        solution: Solution = {
-            ContentId(i): (HostId(primary_mapping[i]), HostId(mirror_mapping[i]))
-            for i in range(self.n_segments)
-        }
+        solution = _pack_solution(primary_mapping, mirror_mapping, self.n_segments)
         
         cost = self._calculate_cost(primary_mapping, mirror_mapping)
 
@@ -1190,10 +1183,7 @@ class LNS:
         self.solver = GreedySolver(config)
         self.rng = random.Random(seed)
         self.n_segments = config.n_segments
-        self.n_hosts = config.n_hosts_target
         self.strategy = config.strategy
-        self.initial_primary_mapping = config.initial_primary_mapping
-        self.initial_mirror_mapping = config.initial_mirror_mapping
         self.printing = printing
 
         self._destroy = LNSDestroyOperators(
@@ -1230,11 +1220,8 @@ class LNS:
         return self.solver.target_primary_load
 
     def solve(self) -> Tuple[Solution, Cost]:
-        ip  = self.config.initial_primary_mapping
-        im  = self.config.initial_mirror_mapping
-
         greedy_sol, _ = self.solver.solve()
-        p, m     = _extract_mapping(greedy_sol, self.n_segments)
+        p, m     = [greedy_sol[i][0] for i in range(self.n_segments)], [greedy_sol[i][1] for i in range(self.n_segments)]
         cost     = self.solver._calculate_cost(p, m)
         best_p, best_m, best_cost = p[:], m[:], cost
 

--- a/gpMgmt/doc/ggrebalance_help
+++ b/gpMgmt/doc/ggrebalance_help
@@ -18,6 +18,7 @@ ggrebalance -x <segment_count>
     [{-D | -S | --no-progress}]
     [--hba-hostnames] [--replay-lag <replay_lag>] [-a]
     [--non-interactive-mode] [-l <log_dir>] [-q] [-v]
+    [--seed <planner_seed>]
 
 ggrebalance
     [-n <parallel_processes>] [-B <batch_size>]
@@ -218,6 +219,10 @@ OPTIONS
  Skip rebalance entirely even if the cluster is unbalanced. If this option is set,
  then ggrebalance does not try to plan rebalance at all. This option makes sense
  only if shrink or expand is needed and the user wants to skip rebalance.
+
+--seed
+ Passing seed value guarantees stable plan generation through several
+ ggrebalance runs. Valid values are any non-negative integer.
 
 -q, --quiet
  Run in quiet mode. Command output is not displayed on

